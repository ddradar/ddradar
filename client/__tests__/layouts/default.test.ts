import {
  createLocalVue,
  mount,
  RouterLinkStub,
  shallowMount,
} from '@vue/test-utils'
import Buefy from 'buefy'

import DefaultLayout from '~/layouts/default.vue'

const localVue = createLocalVue()
localVue.use(Buefy)

describe('/layouts/default.vue', () => {
  const stubs = { NuxtLink: RouterLinkStub, Nuxt: true }
  const $fetchState = { pending: false }
  describe('snapshot test', () => {
    test('renders loading', () => {
      const wrapper = mount(DefaultLayout, {
        localVue,
        stubs,
        mocks: {
          $accessor: {
            isLoggedIn: true,
            name: 'User 1',
            user: { id: 'user_id' },
          },
          $fetchState: { pending: true },
        },
      })
      expect(wrapper).toMatchSnapshot()
    })
    test('renders logout button if authed', () => {
      const wrapper = mount(DefaultLayout, {
        localVue,
        stubs,
        mocks: {
          $accessor: {
            isLoggedIn: true,
            name: 'User 1',
            user: { id: 'user_id' },
          },
          $fetchState,
        },
      })
      expect(wrapper).toMatchSnapshot()
    })
    test('renders login button if not authed', () => {
      const wrapper = mount(DefaultLayout, {
        localVue,
        stubs,
        mocks: { $accessor: { isLoggedIn: false }, $fetchState },
      })
      expect(wrapper).toMatchSnapshot()
    })
  })
  describe('fetch()', () => {
    const fetchUser = jest.fn()
    const $router = { push: jest.fn() }
    beforeEach(() => {
      fetchUser.mockClear()
      $router.push.mockClear()
    })
    test('calls $accessor.fetchUser()', async () => {
      // Arrange
      const $accessor = { fetchUser, auth: true, isLoggedIn: true }
      const wrapper = shallowMount(DefaultLayout, {
        localVue,
        stubs,
        mocks: { $accessor, $fetchState, $router },
      })

      // Act
      await wrapper.vm.$options.fetch?.call(wrapper.vm, null!)

      // Assert
      expect(fetchUser).toBeCalledTimes(1)
      expect($router.push).not.toBeCalled()
    })
    test('redirects /profile if new user', async () => {
      // Arrange
      const $accessor = { fetchUser, auth: true, isLoggedIn: false }
      const wrapper = shallowMount(DefaultLayout, {
        localVue,
        stubs,
        mocks: { $accessor, $fetchState, $router },
      })

      // Act
      await wrapper.vm.$options.fetch?.call(wrapper.vm, null!)

      // Assert
      expect(fetchUser).toBeCalledTimes(1)
      expect($router.push).toBeCalledTimes(1)
      expect($router.push).lastCalledWith('/profile')
    })
  })
})
