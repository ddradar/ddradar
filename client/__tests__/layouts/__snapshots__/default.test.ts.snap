// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`layouts/default.vue renders login button if not authed 1`] = `
<div>
  <nav role="navigation" aria-label="main navigation" class="navbar is-primary">
    <div class="navbar-brand"><a class="navbar-item"><img src="~assets/logo.svg" alt="DDRadar Logo"> <b>DDRadar</b></a><a role="button" aria-label="menu" class="navbar-burger burger"><span aria-hidden="true"></span> <span aria-hidden="true"></span> <span aria-hidden="true"></span></a></div>
    <div class="navbar-menu">
      <div class="navbar-start">
        <div class="navbar-item has-dropdown is-hidden-touch is-hoverable"><a role="menuitem" aria-haspopup="true" href="#" class="navbar-link">SINGLEのレベルから探す</a>
          <div class="navbar-dropdown" style="display: none;"><a class="navbar-item">
              1
            </a><a class="navbar-item">
              2
            </a><a class="navbar-item">
              3
            </a><a class="navbar-item">
              4
            </a><a class="navbar-item">
              5
            </a><a class="navbar-item">
              6
            </a><a class="navbar-item">
              7
            </a><a class="navbar-item">
              8
            </a><a class="navbar-item">
              9
            </a><a class="navbar-item">
              10
            </a><a class="navbar-item">
              11
            </a><a class="navbar-item">
              12
            </a><a class="navbar-item">
              13
            </a><a class="navbar-item">
              14
            </a><a class="navbar-item">
              15
            </a><a class="navbar-item">
              16
            </a><a class="navbar-item">
              17
            </a><a class="navbar-item">
              18
            </a><a class="navbar-item">
              19
            </a></div>
        </div>
        <div class="navbar-item has-dropdown is-hidden-touch is-hoverable"><a role="menuitem" aria-haspopup="true" href="#" class="navbar-link">DOUBLEのレベルから探す</a>
          <div class="navbar-dropdown" style="display: none;"><a class="navbar-item">
              1
            </a><a class="navbar-item">
              2
            </a><a class="navbar-item">
              3
            </a><a class="navbar-item">
              4
            </a><a class="navbar-item">
              5
            </a><a class="navbar-item">
              6
            </a><a class="navbar-item">
              7
            </a><a class="navbar-item">
              8
            </a><a class="navbar-item">
              9
            </a><a class="navbar-item">
              10
            </a><a class="navbar-item">
              11
            </a><a class="navbar-item">
              12
            </a><a class="navbar-item">
              13
            </a><a class="navbar-item">
              14
            </a><a class="navbar-item">
              15
            </a><a class="navbar-item">
              16
            </a><a class="navbar-item">
              17
            </a><a class="navbar-item">
              18
            </a><a class="navbar-item">
              19
            </a></div>
        </div>
        <div class="navbar-item has-dropdown is-hidden-touch is-hoverable"><a role="menuitem" aria-haspopup="true" href="#" class="navbar-link">シリーズから探す</a>
          <div class="navbar-dropdown" style="display: none;"><a class="navbar-item">
              DanceDanceRevolution A20 PLUS
            </a><a class="navbar-item">
              DanceDanceRevolution A20
            </a><a class="navbar-item">
              DanceDanceRevolution A
            </a><a class="navbar-item">
              DanceDanceRevolution (2014)
            </a><a class="navbar-item">
              DanceDanceRevolution (2013)
            </a><a class="navbar-item">
              DDR X3 VS 2ndMIX
            </a><a class="navbar-item">
              DDR X2
            </a><a class="navbar-item">
              DDR X
            </a><a class="navbar-item">
              DDR SuperNOVA2
            </a><a class="navbar-item">
              DDR SuperNOVA
            </a><a class="navbar-item">
              DDR EXTREME
            </a><a class="navbar-item">
              DDRMAX2
            </a><a class="navbar-item">
              DDRMAX
            </a><a class="navbar-item">
              DDR 5thMIX
            </a><a class="navbar-item">
              DDR 4thMIX
            </a><a class="navbar-item">
              DDR 3rdMIX
            </a><a class="navbar-item">
              DDR 2ndMIX
            </a><a class="navbar-item">
              DDR 1st
            </a></div>
        </div>
        <div class="navbar-item has-dropdown is-hidden-touch is-hoverable"><a role="menuitem" aria-haspopup="true" href="#" class="navbar-link">曲名から探す</a>
          <div class="navbar-dropdown" style="display: none;">
            <div class="navbar-item buttons are-small"><a type="button" class="button is-text">
                <!----> <span>
            あ
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            か
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            さ
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            た
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            な
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            は
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            ま
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            や
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            ら
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            わ
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            A
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            B
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            C
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            D
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            E
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            F
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            G
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            H
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            I
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            J
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            K
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            L
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            M
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            N
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            O
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            P
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            Q
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            R
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            S
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            T
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            U
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            V
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            W
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            X
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            Y
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            Z
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            数字・記号
          </span>
                <!----></a></div>
          </div>
        </div>
      </div>
      <div class="navbar-end">
        <div class="navbar-item has-dropdown is-hoverable"><a role="menuitem" aria-haspopup="true" href="#" class="navbar-link">ログイン</a>
          <div class="navbar-dropdown is-right">
            <div class="navbar-item">
              <div class="buttons"><a type="button" href="/.auth/login/twitter" class="button is-info"><span class="icon is-small"><i class="mdi mdi-twitter"></i></span> <span>
              Twitterでログイン
            </span>
                  <!----></a> <a type="button" href="/.auth/login/github" class="button is-dark"><span class="icon is-small"><i class="mdi mdi-github"></i></span> <span>
              GitHubでログイン
            </span>
                  <!----></a></div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </nav>
  <div>
    <transition-stub name="fade">
      <div class="loading-overlay is-active is-full-page">
        <div class="loading-background"></div>
        <div class="loading-icon"></div>
      </div>
    </transition-stub>
    <nuxt-stub></nuxt-stub>
  </div>
  <footer class="footer">
    <div class="content has-text-centered">
      <p>
        不具合を発見した、または新機能の要望がある場合には、
        <a href="https://twitter.com/nogic1008" target="_blank">
          作者のTwitter
        </a>
        または
        <a href="https://github.com/ddradar/ddradar/issues" target="_blank">
          Githubのissue
        </a>
        にてご報告ください。
      </p>
    </div>
  </footer>
</div>
`;

exports[`layouts/default.vue renders logout button if authed 1`] = `
<div>
  <nav role="navigation" aria-label="main navigation" class="navbar is-primary">
    <div class="navbar-brand"><a class="navbar-item"><img src="~assets/logo.svg" alt="DDRadar Logo"> <b>DDRadar</b></a><a role="button" aria-label="menu" class="navbar-burger burger"><span aria-hidden="true"></span> <span aria-hidden="true"></span> <span aria-hidden="true"></span></a></div>
    <div class="navbar-menu">
      <div class="navbar-start">
        <div class="navbar-item has-dropdown is-hidden-touch is-hoverable"><a role="menuitem" aria-haspopup="true" href="#" class="navbar-link">SINGLEのレベルから探す</a>
          <div class="navbar-dropdown" style="display: none;"><a class="navbar-item">
              1
            </a><a class="navbar-item">
              2
            </a><a class="navbar-item">
              3
            </a><a class="navbar-item">
              4
            </a><a class="navbar-item">
              5
            </a><a class="navbar-item">
              6
            </a><a class="navbar-item">
              7
            </a><a class="navbar-item">
              8
            </a><a class="navbar-item">
              9
            </a><a class="navbar-item">
              10
            </a><a class="navbar-item">
              11
            </a><a class="navbar-item">
              12
            </a><a class="navbar-item">
              13
            </a><a class="navbar-item">
              14
            </a><a class="navbar-item">
              15
            </a><a class="navbar-item">
              16
            </a><a class="navbar-item">
              17
            </a><a class="navbar-item">
              18
            </a><a class="navbar-item">
              19
            </a></div>
        </div>
        <div class="navbar-item has-dropdown is-hidden-touch is-hoverable"><a role="menuitem" aria-haspopup="true" href="#" class="navbar-link">DOUBLEのレベルから探す</a>
          <div class="navbar-dropdown" style="display: none;"><a class="navbar-item">
              1
            </a><a class="navbar-item">
              2
            </a><a class="navbar-item">
              3
            </a><a class="navbar-item">
              4
            </a><a class="navbar-item">
              5
            </a><a class="navbar-item">
              6
            </a><a class="navbar-item">
              7
            </a><a class="navbar-item">
              8
            </a><a class="navbar-item">
              9
            </a><a class="navbar-item">
              10
            </a><a class="navbar-item">
              11
            </a><a class="navbar-item">
              12
            </a><a class="navbar-item">
              13
            </a><a class="navbar-item">
              14
            </a><a class="navbar-item">
              15
            </a><a class="navbar-item">
              16
            </a><a class="navbar-item">
              17
            </a><a class="navbar-item">
              18
            </a><a class="navbar-item">
              19
            </a></div>
        </div>
        <div class="navbar-item has-dropdown is-hidden-touch is-hoverable"><a role="menuitem" aria-haspopup="true" href="#" class="navbar-link">シリーズから探す</a>
          <div class="navbar-dropdown" style="display: none;"><a class="navbar-item">
              DanceDanceRevolution A20 PLUS
            </a><a class="navbar-item">
              DanceDanceRevolution A20
            </a><a class="navbar-item">
              DanceDanceRevolution A
            </a><a class="navbar-item">
              DanceDanceRevolution (2014)
            </a><a class="navbar-item">
              DanceDanceRevolution (2013)
            </a><a class="navbar-item">
              DDR X3 VS 2ndMIX
            </a><a class="navbar-item">
              DDR X2
            </a><a class="navbar-item">
              DDR X
            </a><a class="navbar-item">
              DDR SuperNOVA2
            </a><a class="navbar-item">
              DDR SuperNOVA
            </a><a class="navbar-item">
              DDR EXTREME
            </a><a class="navbar-item">
              DDRMAX2
            </a><a class="navbar-item">
              DDRMAX
            </a><a class="navbar-item">
              DDR 5thMIX
            </a><a class="navbar-item">
              DDR 4thMIX
            </a><a class="navbar-item">
              DDR 3rdMIX
            </a><a class="navbar-item">
              DDR 2ndMIX
            </a><a class="navbar-item">
              DDR 1st
            </a></div>
        </div>
        <div class="navbar-item has-dropdown is-hidden-touch is-hoverable"><a role="menuitem" aria-haspopup="true" href="#" class="navbar-link">曲名から探す</a>
          <div class="navbar-dropdown" style="display: none;">
            <div class="navbar-item buttons are-small"><a type="button" class="button is-text">
                <!----> <span>
            あ
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            か
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            さ
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            た
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            な
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            は
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            ま
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            や
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            ら
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            わ
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            A
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            B
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            C
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            D
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            E
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            F
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            G
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            H
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            I
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            J
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            K
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            L
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            M
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            N
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            O
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            P
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            Q
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            R
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            S
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            T
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            U
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            V
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            W
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            X
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            Y
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            Z
          </span>
                <!----></a><a type="button" class="button is-text">
                <!----> <span>
            数字・記号
          </span>
                <!----></a></div>
          </div>
        </div>
      </div>
      <div class="navbar-end">
        <div class="navbar-item has-dropdown is-hoverable"><a role="menuitem" aria-haspopup="true" href="#" class="navbar-link"></a>
          <div class="navbar-dropdown is-right">
            <div class="navbar-item">
              <div class="buttons"><a type="button" class="button is-info"><span class="icon is-small"><i class="mdi mdi-account"></i></span> <span>
              マイページ
            </span>
                  <!----></a> <a type="button" href="/.auth/logout" class="button is-warning">
                  <!----> <span>
              ログアウト
            </span>
                  <!----></a></div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </nav>
  <div>
    <transition-stub name="fade">
      <div class="loading-overlay is-active is-full-page">
        <div class="loading-background"></div>
        <div class="loading-icon"></div>
      </div>
    </transition-stub>
    <nuxt-stub></nuxt-stub>
  </div>
  <footer class="footer">
    <div class="content has-text-centered">
      <p>
        不具合を発見した、または新機能の要望がある場合には、
        <a href="https://twitter.com/nogic1008" target="_blank">
          作者のTwitter
        </a>
        または
        <a href="https://github.com/ddradar/ddradar/issues" target="_blank">
          Githubのissue
        </a>
        にてご報告ください。
      </p>
    </div>
  </footer>
</div>
`;
